refactor: use a error handling similar to CAN

higor-melo opened the pull request:
Depends on:
- [ ] https://github.com/tidewise/drivers-motors_weg_cvw300/pull/14
- [ ] https://github.com/tidewise/drivers-orogen-motors_weg_cvw300/pull/25

# What is this PR for
<!-- A clear description of what this PR do and the changes made.
If you want you can also add the shortcut link here
- [ ] [Shortcut](http://) -->
Implement a CAN Similar Error Handling

# How I did it
<!-- Explain a little bit of your implementation -->

# Results, How I tested
<!-- Explain how you tested you PR, if there is a visual output,
     a GIF or image is welcome -->

# Checklist
- [ ] Assign yourself  in the PR
- [ ] Write unit tests (when relevant)
- [ ] Run rubocop and rake tests locally
- [ ] If this PR initialize a CMAKE package please [enable styling and linting](https://github.com/tidewise/wetpaint-buildconf/blob/master/README.md#enabling-styling-and-linting-checks-for-a-cmake-package)
- [ ] Analyse if this PR modifies the UI, if so:
    - [ ] Tested Tupan's simulation (Charts) :world_map:
    - [ ] Tested Tupan's simulation (Hud) :joystick:
    - [ ] Tested ROV's simulation :joystick:

higor-melo committed:
fix: do error handling similarly to CAN

higor-melo committed:
feat: store validation on it's on method to be used elsewhere

higor-melo committed:
feat: validate packets and retry when needed

doudou commented on src/RTUMaster.cpp:
```diff
@@ -87,91 +97,112 @@ void RTUMaster::readReply(Frame& frame, int function) {
+        throw;
     }
-    while(true);
+    m_error_count += 8;
```
Please avoid magic numbers, and make those numbers configurable. It should be obvious that the overall thresholds and increment/decrement values are very specific to the medium of communication, and that CAN-specific values do not necessarily work for our use-case.

higor-melo replied:
Why should we remove the `while`?

doudou replied:
Not sure whether it is relevant. I did not say anything about removing a while ... ???

doudou commented on src/RTUMaster.cpp:
```diff
@@ -87,91 +97,112 @@ void RTUMaster::readReply(Frame& frame, int function) {
+{
+    if (m_error_count > m_error_threshold) {
+        throw;
     }
```
Can't `throw;` in a non-catch context.

doudou replied:
Refining my comment: this might work, but having a throw without argument in a method feels VERY out of context. Return a boolean from this method, and throw from the `catch` if the method return false.

higor-melo replied:
Done!

doudou reviewed (commented):
Still missing handling of "bad length", which is actually the one error we currently have.

And no unit tests.

doudou commented on src/RTUMaster.cpp:
```diff
@@ -113,21 +113,37 @@ void RTUMaster::writePacketAndReadReply(uint8_t const* buffer,
+            if (m_error_count > m_error_threshold) {
                 throw;
             }
+            increaseErrorCount();
```
Since the error already happened, it would feel more correct to me to increase first and check threshold after.

doudou commented on src/RTUMaster.cpp:
```diff
@@ -113,21 +113,61 @@ void RTUMaster::writePacketAndReadReply(uint8_t const* buffer,
+    int bufsize,
+    Frame& frame,
+    int function,
+    uint8_t const& length)
```
`expected_length`

And do not use const-ref for plain old types.

doudou commented on src/RTUMaster.hpp:
```diff
@@ -52,13 +52,31 @@ namespace modbus {
+            Frame& frame,
+            int function,
+            uint8_t const& length);
+
```
WHY ?

All the packets have an expected length. Just add it to writePacketAndReadReply.

higor-melo replied:
Do they? From what I understood the registers and the digital inputs use the same method, but the comparison is done differently. The Digital input expects the payload to have a `byte_count * 8 < length` instead of a fixed `expected_length`.
Therefore, to avoid using a `lambda` i wrote a new method to be used exclusively by Registers

doudou replied:
Discussion TBD

higor-melo committed:
feat: allow thresholds to be decided

higor-melo commented on src/RTUMaster.hpp:
```diff
@@ -52,12 +52,39 @@ namespace modbus {
+            Frame& frame,
+            int function);
+
+        void writePacketAndReadReplyRegisters(uint8_t const* buffer,
```
Documentation pending

wvmcastro commented on src/RTUMaster.cpp:
```diff
@@ -121,20 +121,44 @@ void RTUMaster::writePacketAndReadReply(uint8_t const* buffer,
+    } while (true);
+}
+
+void RTUMaster::increaseErrorCountOrThrow()
```
Create a custom exception for the error count being above the threshold. It is more readable and you could add a nice description for it

higor-melo replied:
The thing is that by throwing inside the catch context we actually forward throw the error that made the count go above the threshold so it would be either `Unexpected Reply` or `InvalidCRC`. I thought about doing an specific error for it but the idea of not knowing which error actually caused the count to go over, made my go on this direction instead

wvmcastro commented on src/RTUMaster.hpp:
One thing you should think when creating a new functionality on a lib is "how would I guarantee the user uses this safely/correctly?" What I wanna hear from you is how you are obliging the user to set _error increment_ and _error threshold_ so the error handling works as expected?

higor-melo replied:
Nothing TBH, do you suggest creating a method to do this? (that's my ideia)
However i'm not sure where is the best place to place it

wvmcastro replied:
Since these parameters are needed so the object works properly, they must be passed at the constructor and validated there as well

wvmcastro commented on src/RTUMaster.hpp:
```diff
@@ -95,6 +104,24 @@ namespace modbus {
+        uint8_t getErrorThreshold() const;
+
+        /** Get the current error count */
+        uint8_t getErrorCount() const;
```
are all those getters and setter really needed or you just added then no-brainly?

higor-melo replied:
My ideia is to use them to provide feedback. Honestly the one that is actually needed is the `getErrorCount` and the `set*` methods. My question is does it hurt to keep em? My ideia on the orogen would be to report all these values

doudou replied:
> My ideia on the orogen would be to report all these values

The orogen will have to set the parameters values. The only one that needs to be reported is getErrorCount (and, ideally, counters to account separately CRC errors, unexpected replies errors and successful messages).

A.k.a. you will need a stats struct anyways.

wvmcastro reviewed (commented):
[fix: do error handling similarly to CAN](https://github.com/tidewise/drivers-modbus/pull/8/commits/3ed011a0a7ad2f9aedc82158be99398502352984) is actually IMO a `feat`
and [feat: store validation on it's on method to be used elsewhere](https://github.com/tidewise/drivers-modbus/pull/8/commits/3a23d864e7f560db4576c5e1215e6f466c2c95cc)
is a `refactor` or at least a `chore`

doudou commented on src/RTUMaster.hpp:
```diff
@@ -52,13 +52,40 @@ namespace modbus {
+
+        /** Error threshold before throwing */
+        uint8_t m_error_threshold;
+
```
These should be initialized to sane defaults, which would be "try once and fail immediately".

doudou commented on test/test_RTUMaster.cpp:
```diff
@@ -174,15 +185,16 @@ TEST_F(RTUMasterTest, it_accounts_for_invalid_data) {
 }
 
-TEST_F(RTUMasterTest, it_does_timeout_if_the_CRC_error_is_permanent) {
+TEST_F(RTUMasterTest, it_throws_if_the_error_count_reaches_the_max_value) {
```
You are not testing that it is retrying (which was a bug of the original test).

And you now need to test for other problems, like the length problem *which again is what we are trying to solve here*.

higor-melo replied:
We talked this through and it's being tested

higor-melo committed:
feat: allow setting values on initialization

Keep the default constructor as initializing everything with 1 for
threshold and increment

higor-melo committed:
fix: validate registers reply for all functions

higor-melo committed:
feat: report errors through a struct

higor-melo committed:
chore: remove `request` method since it's not used

AFAIK this is not used, I was able to build everything
and it all worked fine. So i'll be removing it and all
the tests related to it.

higor-melo commented:
I'm working on some new unit tests, shouldn't take long though

higor-melo committed:
fix: update total sucess count

add a new test for UnexpectedReply function code and length

doudou commented on src/common.cpp:
```diff
@@ -21,15 +23,12 @@ uint8_t const* common::parse16(uint8_t const* buffer, uint16_t& value)
+                              "size differ (" +
+                              to_string(byte_count + 1u) +
+                              " != " + to_string(frame.payload.size()) + ")");
     }
```
That's now a duplicate from `validateReply`, no ?

higor-melo replied:
Actually it checks the `byte_count` field of the reply however it's redundant. If the `byte_count` is wrong the check against the `expected_value` should also break.
So I'll be removing this

doudou replied:
> So I'll be removing this

I don't understand why ... they are two different pieces of information, so checking both make sense. This checks that the packet is internally consistent, the other checks that it is what the rest of the code expects.

higor-melo replied:
Since the validate reply is now used on the write, we do not have the `byte_count` field on the response. Given that I'll create a new method and validate it again only on the `parse` methods.

doudou commented on src/common.cpp:
```diff
@@ -41,9 +40,9 @@ void common::parseReadDigitalInputs(std::vector<bool>& values,
-        throw UnexpectedReply("RTU::parseReadDigitalInputs: empty reply");
-    }
+    // This is the ammount of bytes needed to send values + 1 for `byte_count`
+    uint8_t byte_length = std::ceil(length / 8.0) + 1;
```
To avoid using floating point calculations here (!)

```
`(length + 7) / 8`
```

doudou commented on src/common.cpp:
```diff
@@ -68,3 +62,17 @@ void common::parseReadDigitalInputs(std::vector<bool>& values,
+    if (frame.payload.empty()) {
+        throw UnexpectedReply("empty reply");
+    }
+
```
This is not necessary given the next check.

doudou commented on src/RTUMaster.cpp:
```diff
@@ -15,11 +16,25 @@ int RTUMaster::extractPacket(uint8_t const* buffer, size_t bufferSize) const
+    setInitialConfig(RTU::FRAME_MAX_SIZE, MAX_PACKET_SIZE);
+}
+
+RTUMaster::RTUMaster(uint8_t error_threshold, uint8_t error_increment)
```
Just add default arguments to both threshold and increment and remove the default constructor. Remember to make the RTUMaster constructor explicit in this case.

doudou replied:
Event without that change, C++11 allows to "reuse" a constructor in another. There's no need for setInitialConfig here.

doudou commented on src/RTUMaster.cpp:
```diff
@@ -111,23 +133,50 @@ vector<uint16_t> RTUMaster::readRegisters(int address,
+            }
+        }
+        catch (UnexpectedReply const&) {
+            m_statistics.total_reply_error_count++;
```
total_unexpected_reply_error_count

doudou commented on src/RTUMaster.cpp:
```diff
@@ -186,7 +239,12 @@ std::vector<bool> RTUMaster::readDigitalInputs(int address,
     auto function = coils ? FUNCTION_READ_COILS : FUNCTION_READ_DIGITAL_INPUTS;
 
-    writePacketAndReadReply(buffer_start, buffer_end - buffer_start, m_frame, function);
+    uint8_t byte_length = std::ceil(count / 8.0) + 1;
```
Same remark on how to compute ceil of an integer division

higor-melo committed:
chore: code improvements

higor-melo committed:
chore: remote include <cmath> and remove floating point

jhonasiv reviewed (commented):
The only thing I didnt check was whether the expected packet length is correct for each method. Since you didnt explicitly change any tests regarding this, I think they are right, but this is not passing CI at moment with errors on motors_weg_cvw300, so maybe not. Waiting for updates on that

higor-melo committed:
fix: the response from the writeSingleRegister is the echo from the message

Therefore the response payload should have 4 bytes of size
corresponding to the Address High, Address Low, Data High,
Data Low

higor-melo commented:
> The only thing I didnt check was whether the expected packet length is correct for each method. Since you didnt explicitly change any tests regarding this, I think they are right, but this is not passing CI at moment with errors on motors_weg_cvw300, so maybe not. Waiting for updates on that

Yep, there was indeed an issue with one of the tests. I'm double checking all the tests using the modbus reference from here: https://www.modbustools.com/modbus.html

seixasxbr commented on src/common.cpp:
```diff
@@ -1,60 +1,40 @@
+    int length)
+{
+    // This is the ammount of bytes needed to send values + 1 for `byte_count`
+    uint8_t byte_length = (length + 7) / 8 + 1;
```
from where this come from?

higor-melo replied:
The request is composed of a starting register address and a amount of registers we want to read, the response however only needs 1 **bit** per input therefore 1 **byte** can provide feedback for 8 addresses and we have to account for the `byte_count`.

Please check the functions 1 and 2: https://www.modbustools.com/modbus.html

higor-melo replied:
I've also took the time to explain it in person. Resolving this

seixasxbr commented on src/common.cpp:
```diff
@@ -1,60 +1,40 @@
-void common::parseReadRegisters(uint16_t* values, Frame const& frame, int length) {
+void common::parseReadRegisters(uint16_t* values, Frame const& frame, int length)
+{
+    validateReply(frame, (length * 2) + 1);
```
from where this come from? 
(length * 2) + 1

higor-melo replied:
Length is the amount of registers we want to read, however the response is composed of 2 bytes (high and low) and the data payload has an extra field called `byte_count`

Please check: https://www.modbustools.com/modbus.html

higor-melo replied:
I've also took the time to explain it in person. Resolving this

seixasxbr commented on src/MasterInterface.hpp:
```diff
@@ -26,11 +26,6 @@ namespace modbus {
          */
         virtual void readReply(Frame& frame, int function) = 0;
 
-        /** Send a request and wait for the slave's reply */
-        virtual Frame const& request(
-            int address, int function, std::vector<uint8_t> const& payload
-        ) = 0;
-
```
why was this removed?

higor-melo replied:
This was not actually used anywhere

seixasxbr commented on src/RTUStatistics.hpp:
```diff
@@ -0,0 +1,30 @@
+namespace modbus {
+    struct RTUStatistics {
+        /** Total InvalidCRC error count */
+        double total_CRC_error_count = 0;
```
```suggestion
        double total_crc_error_count = 0;
```

seixasxbr commented on src/RTUStatistics.hpp:
```diff
@@ -0,0 +1,30 @@
+        uint8_t error_count = 0;
+
+        /** Time of the last update*/
+        base::Time timestamp;
```
```suggestion
        // @meta role logical_time
        base::Time time;
```

seixasxbr commented on src/RTUMaster.cpp:
```diff
@@ -1,75 +1,95 @@
-RTUMaster::RTUMaster()
-    : iodrivers_base::Driver(RTU::FRAME_MAX_SIZE * 10) {
+RTUMaster::RTUMaster(uint8_t error_threshold, uint8_t error_increment)
+    : iodrivers_base::Driver(RTU::FRAME_MAX_SIZE * 10)
```
```suggestion
    : iodrivers_base::Driver(RTU::FRAME_MAX_SIZE * 10)
    , m_error_increment(error_increment)
    , m_error_threshold(error_threshold)
```

seixasxbr commented on src/RTUMaster.cpp:
```diff
@@ -1,75 +1,95 @@
+    m_error_threshold = threshold;
+}
+
+void RTUMaster::setErrorIncrement(uint8_t const& increment)
```
```suggestion
void RTUMaster::setErrorIncrement(uint8_t increment)
```

seixasxbr commented on src/RTUMaster.cpp:
```diff
@@ -1,75 +1,95 @@
 }
 
-Frame RTUMaster::readFrame() {
+void RTUMaster::setErrorThreshold(uint8_t const& threshold)
```
```suggestion
void RTUMaster::setErrorThreshold(uint8_t threshold)
```

pass argument as constant reference when they are non-fundamental types

doudou commented on src/RTUStatistics.hpp:
```diff
@@ -0,0 +1,30 @@
+        double total_unexpected_reply_error_count = 0;
+
+        /** Total Success count */
+        double total_sucess_count = 0;
```
`double` ? To **count** things ?

higor-melo committed:
chore: use int for counting

We're assuming the errors won't happen as much

higor-melo committed:
feat: re-add check against the `byte_count`

higor-melo committed:
chore: improve member initialization

doudou commented on .vscode/c_cpp_properties.json:
```diff
@@ -0,0 +1,20 @@
+{
```
This file should be added to ..gitinore and never committed.

higor-melo replied:
Yeah, it normally is by default. My bad

doudou commented on src/RTUStatistics.hpp:
```diff
@@ -0,0 +1,31 @@
+        int total_unexpected_reply_error_count = 0;
+
+        // Total Success count
+        int total_sucess_count = 0;
```
use fixed-size integer types (e.g. uintXX_t) especially on types that will be put on orogen interfaces.

higor-melo replied:
I went with `uint16_t` for the errors so we assume over long periods of time can be higher the 255 and given that errors are locked to `uint16_t`, i've set the sucess to be the next big thing `uint32_t` to avoid wrap arounds over long operations.
It seems to big to me but it doesn't make sense keeping it the same length as the errors since they are much less likely

higor-melo replied:
we decided to use `uint32_t` and `uint64_t` while keeping the `error_count` as it is

doudou commented on src/RTUStatistics.hpp:
```diff
@@ -0,0 +1,31 @@
+
+        // Time of the last update
+        // @meta role logical_time
+        base::Time time;
```
While not obrigatory, we customarily put the time first.

doudou commented on src/RTUStatistics.hpp:
```diff
@@ -0,0 +1,31 @@
+        uint8_t error_count = 0;
+
+        // Time of the last update
+        // @meta role logical_time
```
```
/** Time of the last update
 *
 * @meta role logical_time
 */
```

higor-melo committed:
chore: define fixed-sized integer types

remove c_cpp_properties.json from git tracked files
typo `sucess` to `success`

doudou commented on src/RTUStatistics.hpp:
```diff
@@ -0,0 +1,33 @@
+         */
+        base::Time time;
+
+        // Total InvalidCRC error count
```
Again.

I can read code, and can infer therefore that `total_crc_error_count` is "the total invalid CRC error count". You need documentation to give more explanations than that. What does this error mean deep down ? What are we counting ?

The doxygen comment format is

```
/** A summary line
 *
 * More detailed explanations if needed. Empty line above important
 */
```

doudou commented on Testing/Temporary/CTestCostData.txt:
```diff
@@ -0,0 +1 @@
+---
```
Also garbage file

doudou commented on Testing/Temporary/LastTest.log:
```diff
@@ -0,0 +1,3 @@
+Start testing: Dec 08 17:00 -03
```
Also garbage file

higor-melo committed:
chore: improve commentary and remove testing files that got added
